### PR TITLE
Add MockL2OutputOracle contract and L2Output Submitter 

### PR DESCRIPTION
**Pull request by cfromknecht**
**Description**
This PR adds a working L2Output batch submitter and integrates it with E2E test harness. To achieve this, a stripped down copy of the EIP-1559 batch submitter is added, and the `l2output` driver is implemented which interacts with ABI bindings for an initial iteration of the `MockL2OutputOracle.sol` contract. The first 5 or so commits could probably be squashed down, though I have left them separate for now to make it clear what was removed from the existing batch submitter. ~There is still some trimming of config options to be done, which I will triage at a later point after getting some feedback on the initial approach.~ For now the batch submitter has it's own `bss` binary, but this can be modified if we prefer a unified binary.

**Metadata**
- Fixes ENG-1928

### CODE REVIEW

**Comment by trianglesphere** [comment]
One large organization question (I only skimmed, didn't dive deep), do we intend to have the `bss` be a separate binary from the `opnode`. It feels like it's exactly copying the structure from the monorepo, but this feels a little odd to have this as a subproject while the `opnode` is a top level project.

**Comment by cfromknecht** [comment]
@trianglesphere I don't have a super strong preference on whether the binaries are separate or the batch submitter is a subcommand. I would lean towards in production running them as separate services in either case though. For now i just left it as a separate binary, since it doesn't impact our ability to integrate with the E2E tests. If there's a strong preference for merging everything into the `opnode` command immediately we can make that happen.

**Comment by tynes** [comment]
I definitely prefer to run them as different binaries so that they hot key can be managed more easily

File: opnode/bss/Makefile
Comment (trianglesphere) [inline opnode/bss/Makefile:10]: Leftover? These otherwise don't appear to be checked in.

File: opnode/bss/config.go
Comment (trianglesphere) [inline opnode/bss/config.go:16]: Splitting the line helps with line length, but makes it hard to grep for the exact error from the error string.

File: opnode/test/system_test.go
Comment (trianglesphere) [inline opnode/test/system_test.go:143]: Could be pulled from the config

File: opnode/test/system_test.go
Comment (trianglesphere) [inline opnode/test/system_test.go:165]: `endpoint(cfg.l1.nodeConfig)` returns the WS endpoint. It could be renamed to `wsEndpoint` and a `httpEndpoint(...)` could be added if the batch submitter prefers using HTTP

File: opnode/bss/package.json
Comment (trianglesphere) [inline opnode/bss/package.json:2]: Do we have docs on this/the changelog?

**Review by trianglesphere — COMMENTED** [review]
Thanks for clarifying that the bss will be it's own binary/service. With that it feels like the directory structure has it as a submodule rather than a peer to the deposit sequencer node.

I know this was copied from the mono repo, but I wonder if the bss needs it's own `go.mod` file - i.e. will it be versioned independently from the rest of the rollup node components.

File: opnode/bss/config.go
Comment (tynes) [reply opnode/bss/config.go:139]: This is how we configure the existing batch submitter. Using an encrypted keystore (https://github.com/roynalnaruto/eth-keystore-rs) could help to improve our security practices

File: opnode/bss/crypto.go
Comment (tynes) [reply opnode/bss/crypto.go:9]: Perhaps https://github.com/miguelmota/go-ethereum-hdwallet ?

File: opnode/bss/config.go
Comment (tynes) [reply opnode/bss/config.go:126]: These config options were used at one point with the old batch submitter

File: opnode/bss/config.go
Comment (tynes) [reply opnode/bss/config.go:139]: I wish that encrypted keystores worked with mnemonics by holding information about the hd path as part of its datastructure. Or if the encrypted keystore could hold a mnemonic and then the app could choose to derive whatever key it wants

File: opnode/bss/crypto.go
Comment (cfromknecht) [reply opnode/bss/crypto.go:9]: `go-ethereum-hdwallet` is just a wrapper around `btcutil/hdkeychain` for the things we use it for, but swapping it out did allow us to get rid of the `crypto.go` and `crypto_test.go` files 👍

File: opnode/bss/CHANGELOG.md
Comment (cfromknecht) [reply opnode/bss/CHANGELOG.md]: removed the changelog, we can add it if we decide we want two separate services

File: opnode/bss/config.go
Comment (cfromknecht) [reply opnode/bss/config.go]: yes this config option is used in various test environments in the monorepo. i went ahead and removed the option to set private keys explicitly since this test environment already uses mnemonics

File: opnode/bss/config.go
Comment (cfromknecht) [reply opnode/bss/config.go]: > Why do we import a mnemonic and specify derivation paths? I think we should instead specify a file-path towards an encrypted keystore and another to a keystore password, and mount/protect that password file with some key vault when deployed in production. IMHO mnemonics should only ever be used outside of deployed code, for backups etc. and never for key inputs.

I don't know the specifics of how it's implemented but in prod we use Google's secret manager to inject these. I have to imagine under the hood the setup is fairly similar to what you describe here—the mnemonic is not stored in a config file if that's your concern. As far using a raw key vs mnemonic+hdpath I don't have a strong preference. In the current BSS we use the mnemonic since it allows us to derive a key for each service (tx and state-roots) using only 1 pice of secret data, rather than multiple.

File: opnode/bss/Makefile
Comment (cfromknecht) [reply opnode/bss/Makefile]: Yes good catch, removed.

File: opnode/bss/config.go
Comment (cfromknecht) [reply opnode/bss/config.go]: This error has been removed, but i did a quick grep and it seems like all of the other instances are on one line 👍

File: opnode/test/system_test.go
Comment (cfromknecht) [reply opnode/test/system_test.go]: done

File: opnode/test/system_test.go
Comment (cfromknecht) [reply opnode/test/system_test.go]: i hadn't thought to try the bss with WS's, but it seems to work just fine. removed the `httpHost` logic

File: opnode/bss/package.json
Comment (cfromknecht) [reply opnode/bss/package.json]: I've removed the changelog for now, we can add it at a later time if we want

**Comment by cfromknecht** [comment]
@protolambda @trianglesphere thanks for the reviews! I was able to whittle down the remainder of the config options and code down to a more minimal set.

The main thing that comes to mind rn—are we okay with leaving the `bss` binary separate? Or should I go ahead and make it a subcommand? The main thing here is that we currently use two different libraries for config parsing (`ask` for `opnode`, and `urfave/cli` for `bss`). I'd like to solidify on around one before proceeding.

@trianglesphere I moved the batch submitter from `opnode/bss` to `bss` given your prior comment, and also removed the `bss/go.mod` so that it is part of the top-level module.

**Comment by trianglesphere** [comment]
> The main thing that comes to mind rn—are we okay with leaving the `bss` binary separate? Or should I go ahead and make it a subcommand? The main thing here is that we currently use two different libraries for config parsing (`ask` for `opnode`, and `urfave/cli` for `bss`). I'd like to solidify around one before proceeding.

I think separate binaries are fine for now. Given that they are planned to be ran separately I don't see a large benefit in unifying them under one command.

**Comment by trianglesphere** [comment]
Completely separately (and not something that should be solved in this PR), have we considered making the tx manager a separate service (like it's own process) that we then connect to (probably need some sort of authentication).

Another idea here is to pull out the tx manager to a shared package - it seems like a good candidate given how much we use it from different contexts.

**Review by protolambda — APPROVED** [review]
Over time I definitely think we should use encrypted keystores as opposed to mnemonics (best security practice, don't want a leak to leak all other account private keys derivable from a mnemonic), but otherwise this all looks good.

**Review by norswap — APPROVED** [review]
Interesting stuff!

Made some minor comments (typos, etc), and asked a few questions where the logic is not entirely clear to me + some high level items.

Would be interested to know what the future work on this is. Needs an issue to track it if this is merged.

**Comment by norswap** [comment]
Related to #170 (epic: #156)

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (maurelian) [inline packages/contracts/contracts/L1/StateRootOracle.sol:59]:
```suggestion
    function deleteState(uint _timestamp) external {
```

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (maurelian) [inline packages/contracts/contracts/L1/StateRootOracle.sol:8]: @cfromknecht for the purpose of this PR, would it be possible to add just the interface for this contract, or the minimum viable amount of contract you need to implement the go-code?

**Review by maurelian — CHANGES_REQUESTED** [review]
I'd prefer not to merge a new contract without testing. If necessary for the sake of implementing and testing the go code, we could create a Mock instead.

File: bss/flags/flags.go
Comment (cfromknecht) [reply bss/flags/flags.go:44]: the `cli.DurationFlag` parses inputs of the form `1s`, `500ms`, `2h`, there is no inherent unit of time

File: bss/txmgr/txmgr.go
Comment (cfromknecht) [reply bss/txmgr/txmgr.go:198]: A tx sent by a prior running instance confirming is the most likely.

> Also, can't this be handled by bumping up the nonce and retrying?

Yes, that's exactly what happens. The submission at the current nonce is aborted, and the system reevaluates what needs to be done at the next nonce and if necessary calls back into `Send`. We don't want to blindly bump the nonce here as the confirming txn may contain the same data.

File: bss/txmgr/txmgr.go
Comment (cfromknecht) [reply bss/txmgr/txmgr.go:225]: Correct. If this fails while reorging we will just sleep and wait to try again

File: bss/Makefile
Comment (cfromknecht) [reply bss/Makefile:3]: The `GITVERSION` name is a little confusing. This should actually just be `VERSION`. In prod this is pulled from the `package.json` which is updated by `changeset`, but afaict we don't have that set up in this repo so I just made a placeholder.

File: bss/flags/flags.go
Comment (cfromknecht) [reply bss/flags/flags.go]: turns out we don't use this flag, removed

File: bss/config.go
Comment (cfromknecht) [reply bss/config.go:21]: oops, the `SCC` part was left over from when this was renamed from `SCCAddress`. i've replaced `SCC` with `SRO` since this matches the current name of the solidity file, but we can definitely change it when we agree on a new name :)

File: bss/config.go
Comment (cfromknecht) [reply bss/config.go]: I ended up removing `ValidateConfig` altogether, it's really not as useful here as it is in the production BSS where we do some other sanity checks.

File: bss/batch_submitter.go
Comment (cfromknecht) [reply bss/batch_submitter.go]: yes this will block forever. it's not super critical to add signal handling since we don't have really have anything that needs proper closing, i.e. a database connection, but added it anyways just in case 👍

File: bss/batch_submitter.go
Comment (cfromknecht) [reply bss/batch_submitter.go]: done!

File: bss/drivers/l2output/driver.go
Comment (cfromknecht) [reply bss/drivers/l2output/driver.go]: fixed

File: bss/drivers/l2output/driver.go
Comment (cfromknecht) [reply bss/drivers/l2output/driver.go]: I don't think so, if we're executing this logic that means the prior submission happened 25 minutes/100 blocks in the past, which would be a very deep reorg. I changed it to return an error just in case tho, which would be safer.

File: bss/drivers/l2output/driver.go
Comment (cfromknecht) [reply bss/drivers/l2output/driver.go:197]: I worked the logging slightly here to call it a checkpoint rather than a batch, which I think works well. Happy to change the terminology further

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (cfromknecht) [reply packages/contracts/contracts/L1/StateRootOracle.sol]: removed this method 👍

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (cfromknecht) [reply packages/contracts/contracts/L1/StateRootOracle.sol:7]: No problem, I stripped down the contract to get to an MVP 👍

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (cfromknecht) [reply packages/contracts/contracts/L1/StateRootOracle.sol:1]: > This file has mixed tabs and spaces, and has a lot of indentation / formatting issue. I'll make a commit to fix those (feel free to revert it if you made concurrent changes to this file).

Realized this was an outdated version of the contract, should be better now.

> Also we probably should rename "state root" into "l2 output commitment".

So `L2OutputCommitmentOracle`? Could also be something like `L2OutputCheckpointOracle` or`L2CheckpointOracle`

**Comment by cfromknecht** [comment]
> I'd prefer not to merge a new contract without testing. If necessary for the sake of implementing and testing the go code, we could create a Mock instead.

@maurelian I've never written solidity tests before but I can give it a shot. The contract has some coverage through the e2e test, which might be an easier place to expand for the sake of this PR

File: bss/bindings/sro/state_root_oracle.go
Comment (maurelian) [inline bss/bindings/sro/state_root_oracle.go:1]: Might be a naive Go question, but should this be checked-in, or generated by the build?

Comment (trianglesphere) [reply bss/bindings/sro/state_root_oracle.go:1]: Generally we'd like to have these checked in.

**Comment by maurelian** [comment]
I too hate when someone just drives by, leaves a few comments, then requests changes without clearly stating what it would take to get their approval. So I'll clarify/enumerate what I'd like to see prior to approving this PR: 

1. [ ] Rename the contract by prefixing it with 'Mock' and run 'yarn lint' on it.  
  - It's untested and full of holes (ie. no auth on `deleteStatebatch`). I'd prefer to even reduce the functionality to a mock, but a renaming to make it clear we're not merging broken reference code would be OK with me. 
2. [ ] Update names from the "Batch Submission" paradigm to match with the "L2 Output Submission"  paradigm.

I'd also _prefer_ to have specs in before ref impl, but it's kind of a chicken and egg thing, so I don't want to be overly pedantic. That said, it just means that when the spec (#194) comes in, it will need to be reviewed to ensure it matches the impl.

File: bss/bindings/sro/state_root_oracle.go
Comment (protolambda) [reply bss/bindings/sro/state_root_oracle.go:1]: Generated Go code should be checked in. We can use the `go:generate` thing to automate re-generating it, and we should add something to CI to re-generate everything and diff with the existing files, to make sure the generated code is correct.

Comment (trianglesphere) [reply bss/bindings/sro/state_root_oracle.go:1]: FYI a lot of these are generated with `make` (easier to handle with the output of `hardhat`), but adding a CI step to check that the generated files are up to date would be a nice to have.

**Comment by cfromknecht** [comment]
> Rename the contract by prefixing it with 'Mock' and run 'yarn lint' on it.

Done!

> Update names from the "Batch Submission" paradigm to match with the "L2 Output Submission" paradigm.

@maurelian I've renamed the top level `BatchSubmitter` struct to `L2OutputSubmitter`, updated log lines, help msg, etc., and removed all reference of batches in comments favor of L2Outputs. @norswap I've renamed the package `l2os` for L2 Output Submitter, but if you have a better name I can change it again.

I do wonder how future proof naming the top-level struct `L2OutputSubmitter` is—once we add logic in to submit tx batches it won't be quite accurate. I wonder if there's a better name that would be more generic, but we can also deal with that bridge when we get there as it is accurate for the time being.

**Review by maurelian — APPROVED** [review]

**Review by maurelian — APPROVED** [review]

File: bss/flags/flags.go
Comment (norswap) [reply bss/flags/flags.go]: Is that a common Go thing? Otherwise might be worth mentionning that you can pass the unit. What if you don't mention a unit?

File: bss/Makefile
Comment (norswap) [reply bss/Makefile]: Could you make a comment to that effect?

File: packages/contracts/contracts/L1/StateRootOracle.sol
Comment (norswap) [reply packages/contracts/contracts/L1/StateRootOracle.sol]: All of these are fine by me, I guess "output commitment" is what is predominantly used in the code? (I don't remember!)

File: bss/drivers/l2output/driver.go
Comment (norswap) [reply bss/drivers/l2output/driver.go]: So a checkpoint will be batched transactions + output commitment, with the transaction part not yet implemented, is that the idea?

**Review by tynes — APPROVED** [review]

File: bss/flags/flags.go
Comment (cfromknecht) [reply bss/flags/flags.go]: yes it’s pretty common for go. under the hood urfave just calls into the stdlib `time.ParseDuration` https://pkg.go.dev/time#ParseDuration

if no units are provided it defaults to ns

**Comment by cfromknecht** [comment]
I'd like to suggest holding off on merging until https://github.com/ethereum-optimism/optimism/pull/2165 lands, that way I can update this PR with the fix. Also another reason to make a standalone `txmgr` package..

**Comment by maurelian** [comment]
> I'd like to suggest holding off on merging until [ethereum-optimism/optimism#2165](https://github.com/ethereum-optimism/optimism/pull/2165) lands, that way I can update this PR with the fix. Also another reason to make a standalone `txmgr` package..

OK, setting to draft.

File: l2os/cmd/l2os/main.go
Comment (trianglesphere) [inline l2os/cmd/l2os/main.go:33]: FYI this pulls from geth's hardcoded version.

**Comment by maurelian** [comment]
@cfromknecht I see that https://github.com/ethereum-optimism/optimism/pull/2165 has been merged. 
What does that mean for the status of this PR? Does it need to incorporate changes from there?

File: l2os/cmd/l2os/main.go
Comment (trianglesphere) [reply l2os/cmd/l2os/main.go:33]: yep

Comment (cfromknecht) [reply l2os/cmd/l2os/main.go:33]: Fixed, just used a simple `<version>-<commit>-<date>` format

**Comment by mslipper** [comment]
I believe this is ready for merge.

**Comment by cfromknecht** [comment]
> I believe this is ready for merge.

Rebased!

**Comment by maurelian** [comment]
I'm hesitant to merge this large PR given the gaps in our review process we've recently discussed. So I've asked @trianglesphere to give it another review, but first I'd appreciate if @cfromknecht could add the following in the way of details:

1. What are the security risks in this code and how are they mitigated? 
2. Which areas of the code require the most scrutiny?

**Review by trianglesphere — APPROVED** [review]
I think we should go ahead and merge this.

Looking through the code, I was looking for specs, Frequency/the exact output, and internal states that can get messed up.

My impression is that we need to follow up with specs on how often we should commit to the output. This is based on my experience, but I think we rely too much on timestamps when determining when we should have an output commitment.

Some other thoughts on specs (more from the side of verifying the output commitment)
- Over what range is the commitment
- How delayed can the commitment be from when it is available to be committed to and when it must be on chain (I guess there is an economic staking component to this) (or can anyone submit an update after a while)
- How do withdrawals interact with having outputs (and not having them)
- Do we want to tie output commitments to a specific chain (i.e. include a parameter for the parent block hash (if this can be done))

**Comment by trianglesphere** [comment]
@cfromknecht 🚢 it

**Comment by JahJah01** [comment]
Thax for the merge

**Comment by JahJah01** [comment]
Thax for the merge it was a grade experience

**Comment by JahJah01** [comment]
Thax merge

**Comment by JahJah01** [comment]
Thak you

**Pull request merged**
